app/integration_action.go

```diff
@@ -250,6 +257,87 @@ func (a *App) DoActionRequest(rawURL string, body []byte) (*http.Response, *mode
 	return resp, nil
 }
 
+type LocalResponseWriter struct {
+	data   []byte
+	status int
+}
+
+func (w *LocalResponseWriter) Header() http.Header {
+	return make(http.Header)
+}
+
+func (w *LocalResponseWriter) Write(bytes []byte) (int, error) {
+	w.data = make([]byte, len(bytes))
+	copy(w.data, bytes)
+	return len(w.data), nil
+}
+
+func (w *LocalResponseWriter) WriteHeader(statusCode int) {
+	w.status = statusCode
+}
+
+func (a *App) DoLocalRequest(rawURL string, body []byte) (*http.Response, *model.AppError) {
+	inURL, err := url.Parse(rawURL)
+	if err != nil {
+		return nil, model.NewAppError("DoActionRequest", "api.post.do_action.action_integration.app_error", nil, "err="+err.Error(), http.StatusBadRequest)
+	}
+	pluginsEnvironment := a.GetPluginsEnvironment()
+
+	regx := regexp.MustCompile(`plugins\/\s*([^\/]*)`)
+	matches := regx.FindStringSubmatch(inURL.Path)
+	if len(matches) != 2 {
+		return nil, model.NewAppError("DoActionRequest", "api.post.do_action.action_integration.app_error", nil, "err=Unable to find pluginId", http.StatusBadRequest)
+	}
+	pluginId := matches[1]
+
+	regx = regexp.MustCompile(`(` + pluginId + `)(\/.*)`)
+	matches = regx.FindStringSubmatch(inURL.Path)
+	if len(matches) != 3 {
+		return nil, model.NewAppError("DoActionRequest", "api.post.do_action.action_integration.app_error", nil, "err=Unable to find path", http.StatusBadRequest)
+	}
+	path := matches[2]
+
+	hooks, err := pluginsEnvironment.HooksForPlugin(pluginId)
+	if err != nil {
+		return nil, model.NewAppError("DoActionRequest", "api.post.do_action.action_integration.app_error", nil, "err="+err.Error(), http.StatusBadRequest)
```

Comment:
I'm thinking we need a 404 for the case where the plugin doesn't exist, and an internal server error otherwise.

Reply:
👍